### ISSUE TITLE
Close /dev/null in twisted.internet.process._FDDetector._checkDevFDSanity explicitly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5008]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5008|
|Type|enhancement|
|Created|2011-03-28 13:34:51Z|

The current implementation lets reference counting/garbage collection close it, which may result in it being left open for longer than expected or desired.

A patch is attached to [#4881](https://github.com/twisted/twisted/issues/4881) which addresses this.

<details><summary>Searchable metadata</summary>

```
trac-id__5008 5008
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1301319291000000 1301319291000000
changetime__1302214621000000 1302214621000000
version__None None
owner__mwh mwh

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_1"></a>|@mwhudson commented|
|-|-|

(In [31553]) Branching to 'close-dev-null-5008'

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_2"></a>|@mwhudson commented|
|-|-|

I applied the patch in [31555]

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @mwhudson|
|-|-|

I think one of the tests should assert that the fake file is closed.  Otherwise, the change just _lets_ the implementation close it, but doesn't verify that it does.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_4"></a>|@mwhudson removed owner|
|-|-|

Fair enough.  I added a test in [31559].

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/close-dev-null-5008)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @mwhudson|
|-|-|

Looks good, please merge.  Thanks

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_7"></a>|@mwhudson set status to `closed`|
|-|-|

(In [31569]) Merge close-dev-null-5008: explicitly close /dev/null in t.i.process._FDDetector._checkDevFDSanity

Author: lewq, mwh
Reviewer: exarkun
Fixes: [#5008](https://github.com/twisted/twisted/issues/5008)

twisted.internet.process._FDDetector._checkDevFDSanity now explicitly closes
the file descriptor it opens rather than just waiting for GC to close it.
